### PR TITLE
Python Storage

### PR DESCRIPTION
**Pull request by yimingnzhao**
Implemented storage of Gensim objects via Google Cloud Storage.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -3,3 +3,4 @@ src/main/webapp/js/config.js
 src/main/webapp/WEB-INF/appengine-web.xml
 python/env/*
 __pycache__/
+python/static/resources/lecture-buddy-service.json
```
Comment (RickyAvina) [inline .gitignore:6]: This is the service downloaded from the console

File: python/app.yaml [hunk 2]
```diff
@@ -1,7 +1,7 @@
 runtime: python37
 service: python
 
-instance_class: F4
+instance_class: F4_1G
```
Comment (RickyAvina) [inline python/app.yaml:4]: Reference: https://cloud.google.com/appengine/docs/standard#instance_classes

File: src/main/webapp/js/captions.js [hunk 3]
```diff
@@ -100,34 +105,84 @@ function execute(url) {
...
-            console.log("Fetching captions from database...");
-        } else {
-            // video id not found in db, fetching from Youtube API
-            getTrackId(videoId)
-            .then(trackId => getYTCaptions(trackId))
-            .then(captions => parseCaptionsIntoJson(captions, url))
-            .then(parsed_captions => sendJsonForm(parsed_captions))
-            .then(nlp_json => successfulDisplay(nlp_json));
-        }
+function callGensim(video_id, yt_captions) {
```
Comment (yimingnzhao) [inline src/main/webapp/js/captions.js:137]: Add function description

File: src/main/webapp/js/captions.js [hunk 4]
```diff
@@ -100,34 +105,84 @@ function execute(url) {
...
+                        .then(captions => parseCaptionsIntoJson(captions, $('#search-wrapper input').val()))
+                        .then(parsed_captions => sendJsonForm(parsed_captions))
+                        .then(nlp_json => successfulDisplay(nlp_json));
+                    }
+                });
+            });
+}
+
+
+function getFullCaptionsText(timeRangedText) {
```
Comment (yimingnzhao) [inline src/main/webapp/js/captions.js:180]: ditto

File: src/main/webapp/js/captions.js (hunk 3 above)
Comment (yimingnzhao) [reply src/main/webapp/js/captions.js:142]: done

File: src/main/webapp/js/captions.js (hunk 4 above)
Comment (yimingnzhao) [reply src/main/webapp/js/captions.js:190]: done

File: src/main/webapp/js/captions.js [hunk 5]
```diff
@@ -100,34 +105,94 @@ function execute(url) {
...
-            .then(captions => parseCaptionsIntoJson(captions, url))
-            .then(parsed_captions => sendJsonForm(parsed_captions))
-            .then(nlp_json => successfulDisplay(nlp_json));
-        }
+/**
+ * Calls the gensim server to set up the current video id and captions for processing
+ * @param video_id the string id of the YouTube video
+ * @param yt_captions the list of yt_captions
+ */
+function callGensim(video_id, yt_captions) {
```
Comment (jiashenwang) [inline src/main/webapp/js/captions.js:142]: The name here is not that meaningful. Plz change it to sth like processVideoOnGensim

File: src/main/webapp/js/captions.js [hunk 6]
```diff
@@ -100,34 +105,94 @@ function execute(url) {
...
+            });
+}
+
+
+/**
+ * Gets the full text from a list of TimeRangedText objects
+ * @param timeRangedText the list of TimeRangedText objects
+ * @return the string of the full captions text
+ */
+function getFullCaptionsText(timeRangedText) {
```
Comment (jiashenwang) [inline src/main/webapp/js/captions.js:190]: getFullCaptionsFromTimeRangedText

File: src/main/webapp/js/captions.js [hunk 7]
```diff
@@ -624,17 +693,26 @@ function setCaptionsButton() {
...
-                    // Sets the results table
-                    document.getElementById("FullCap").innerHTML = text;
-                }
-            })
-            .catch(err => renderError(err));
+            if (fullCaptions == '') {
+                fetch('/fullcaption?id=' + currentVideoID, {
+                    method: 'GET',
+                })
+                    .then((response) => response.json())
```
Comment (jiashenwang) [inline src/main/webapp/js/captions.js:700]: Could we combine these 2 then here?

File: src/main/webapp/js/captions.js [hunk 8]
```diff
@@ -624,17 +693,26 @@ function setCaptionsButton() {
...
+                    .then ((list) => {
+                    if (list != null) {
```
Comment (jiashenwang) [inline src/main/webapp/js/captions.js:702]: Let's try to reduce the scope
```
if (list != null) {
    handlerError();
    return;
}
```

File: src/main/webapp/js/captions.js [hunk 9]
```diff
@@ -624,17 +693,26 @@ function setCaptionsButton() {
...
-            })
-            .then((response) => response.text())
-            .then ((text) => {
-                if (text != null && text.trim() != '') {
-                    // Sets the results table
-                    document.getElementById("FullCap").innerHTML = text;
-                }
-            })
-            .catch(err => renderError(err));
+            if (fullCaptions == '') {
```
Comment (jiashenwang) [inline src/main/webapp/js/captions.js:696]: Ditto
```
if (1$('#FullCap').is(':empty')) {
    $('#FullCap').empty();
    return;
}
if (fullCaptions != '') {
    $('#FullCap').text(fullCaptions.replace(/\[.*?\] /g, ''));
    return;
}
...
```

File: src/main/webapp/js/gensim.js [hunk 10]
```diff
@@ -16,15 +17,50 @@ async function postGensim(url, data, query, callback=handleGensimResponse) {
...
+            body: JSON.stringify({'data': data,
+                                'v_id': v_id})
+        }).catch(error => { throw error });
 
         callback(response);
-        // return response;
+
+    } catch(err) {
+        console.log(err);
+        $('#query-output table').empty();
```
Comment (jiashenwang) [inline src/main/webapp/js/gensim.js:28]: Line 28 - 31 seems repeated, could we create a function handleError to do the UI change?

File: src/main/webapp/js/captions.js (hunk 5 above)
Comment (yimingnzhao) [reply src/main/webapp/js/captions.js]: addressed

File: src/main/webapp/js/captions.js (hunk 6 above)
Comment (yimingnzhao) [reply src/main/webapp/js/captions.js]: addressed

File: src/main/webapp/js/captions.js (hunk 7 above)
Comment (yimingnzhao) [reply src/main/webapp/js/captions.js]: addressed

File: src/main/webapp/js/captions.js (hunk 9 above)
Comment (yimingnzhao) [reply src/main/webapp/js/captions.js:717]: addressed

**Pull request merged**